Support RRef.backward() for User RRefs.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46641 Support RRef.backward() for User RRefs.**

Second part of https://github.com/pytorch/pytorch/pull/46568, allows
RRef.backward() to work for user RRefs.

Differential Revision: [D24441300](https://our.internmc.facebook.com/intern/diff/D24441300/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24441300/)!

pritamdamania committed:
Support RRef.backward() for Owner RRefs.

Second part of https://github.com/pytorch/pytorch/pull/46568, allows
RRef.backward() to work for owner RRefs.

Differential Revision: [D24441300](https://our.internmc.facebook.com/intern/diff/D24441300/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24441300/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Support RRef.backward() for Owner RRefs."


Second part of https://github.com/pytorch/pytorch/pull/46568, allows
RRef.backward() to work for owner RRefs.

Differential Revision: [D24441300](https://our.internmc.facebook.com/intern/diff/D24441300/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24441300/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Support RRef.backward() for Owner RRefs."


Second part of https://github.com/pytorch/pytorch/pull/46568, allows
RRef.backward() to work for owner RRefs.

Differential Revision: [D24441300](https://our.internmc.facebook.com/intern/diff/D24441300/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24441300/)!

[ghstack-poisoned]

mrshenli reviewed (commented):
A general question: what's the reason for adding two new message types instead of using a helper Python function on top of RPC?

pritamdamania87 commented:
> A general question: what's the reason for adding two new message types instead of using a helper Python function on top of RPC?

I felt this would be easier to torchscript in the future if needed? Plus since PyRRef is completely defined in C++ land, I probably might have to do some messy `setattr` to extend PyRRef with a new method called "backward".

pritamdamania committed:
Update on "Support RRef.backward() for User RRefs."


Second part of https://github.com/pytorch/pytorch/pull/46568, allows
RRef.backward() to work for user RRefs.

Differential Revision: [D24441300](https://our.internmc.facebook.com/intern/diff/D24441300/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24441300/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Support RRef.backward() for User RRefs."


Second part of https://github.com/pytorch/pytorch/pull/46568, allows
RRef.backward() to work for user RRefs.

Differential Revision: [D24441300](https://our.internmc.facebook.com/intern/diff/D24441300/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24441300/)!

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/autograd/rpc_messages/rref_backward_req.h:
```diff
@@ -0,0 +1,39 @@
+      int64_t autogradContextId,
+      bool retainGraph = false);
+
+  const rpc::RRefId& getRRefId();
+
+  int64_t getAutogradContextId();
+
+  bool retainGraph();
```
these function can all be `const`?

mrshenli commented on torch/csrc/distributed/autograd/rpc_messages/rref_backward_req.h:
```diff
@@ -0,0 +1,39 @@
+      const rpc::Message& message);
+
+ private:
+  rpc::RRefId rrefId_;
+  int64_t autogradContextId_;
+  bool retainGraph_;
```
They can be const vars?

mrshenli commented on torch/csrc/distributed/autograd/rpc_messages/rref_backward_resp.cpp:
```diff
@@ -0,0 +1,18 @@
+
+std::unique_ptr<RRefBackwardResp> RRefBackwardResp::fromMessage(
+    const rpc::Message& message) {
+  return std::unique_ptr<RRefBackwardResp>();
```
do we need to check the message type matches?

mrshenli commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -285,22 +286,56 @@ c10::IValue PyRRef::toIValue() const {
+  backward(autogradContextId, retainGraph, rref_);
+}
+
+void PyRRef::backward(
```
curious, what's the reason for adding this static method to PyRRef  instead of adding an instance method backward to RRef?

pritamdamania87 replied:
I didn't want to add a Python dependency to RRef.

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -3684,6 +3684,29 @@ def test_local_rref_backward(self):
+        with dist_autograd.context() as context_id:
+            rref = rpc.remote(dst, RpcTest._identity, args=("foo",))
+            with self.assertRaisesRegex(RuntimeError, "RRef should contain a tensor for .backward()"):
+                rref.backward(context_id)
```
adding a test for calling backward without providing an `ctx_id`?

mrshenli reviewed (approved):
LGTM! added some minor comments

pritamdamania committed:
Update on "Support RRef.backward() for User RRefs."


Second part of https://github.com/pytorch/pytorch/pull/46568, allows
RRef.backward() to work for user RRefs.

Differential Revision: [D24441300](https://our.internmc.facebook.com/intern/diff/D24441300/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24441300/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Support RRef.backward() for User RRefs."


Second part of https://github.com/pytorch/pytorch/pull/46568, allows
RRef.backward() to work for user RRefs.

Differential Revision: [D24441300](https://our.internmc.facebook.com/intern/diff/D24441300/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24441300/)!

[ghstack-poisoned]
